newton/solvers/solver_featherstone.py

```diff
@@ -0,0 +1,2112 @@
+
+
+@wp.kernel
+def integrate_generalized_joints(
+    joint_type: wp.array(dtype=int),
+    joint_q_start: wp.array(dtype=int),
+    joint_qd_start: wp.array(dtype=int),
+    joint_axis_dim: wp.array(dtype=int, ndim=2),
+    joint_q: wp.array(dtype=float),
+    joint_qd: wp.array(dtype=float),
+    joint_qdd: wp.array(dtype=float),
+    dt: float,
+    # outputs
+    joint_q_new: wp.array(dtype=float),
+    joint_qd_new: wp.array(dtype=float),
+):
+    # one thread per-articulation
+    index = wp.tid()
+
+    type = joint_type[index]
+    coord_start = joint_q_start[index]
+    dof_start = joint_qd_start[index]
+    lin_axis_count = joint_axis_dim[index, 0]
+    ang_axis_count = joint_axis_dim[index, 1]
+
+    jcalc_integrate(
+        type,
+        joint_q,
+        joint_qd,
+        joint_qdd,
+        coord_start,
+        dof_start,
+        lin_axis_count,
+        ang_axis_count,
+        dt,
+        joint_q_new,
+        joint_qd_new,
+    )
+
+
+class FeatherstoneSolver(SolverBase):
+    """A semi-implicit integrator using symplectic Euler that operates
+    on reduced (also called generalized) coordinates to simulate articulated rigid body dynamics
+    based on Featherstone's composite rigid body algorithm (CRBA).
+
+    See: Featherstone, Roy. Rigid Body Dynamics Algorithms. Springer US, 2014.
+
+    Instead of maximal coordinates :attr:`State.body_q` (rigid body positions) and :attr:`State.body_qd`
+    (rigid body velocities) as is the case :class:`SemiImplicitIntegrator`, :class:`FeatherstoneIntegrator`
+    uses :attr:`State.joint_q` and :attr:`State.joint_qd` to represent the positions and velocities of
+    joints without allowing any redundant degrees of freedom.
+
+    After constructing :class:`Model` and :class:`State` objects this time-integrator
+    may be used to advance the simulation state forward in time.
+
+    Note:
+        Unlike :class:`SemiImplicitIntegrator` and :class:`XPBDIntegrator`, :class:`FeatherstoneIntegrator` does not simulate rigid bodies with nonzero mass as floating bodies if they are not connected through any joints. Floating-base systems require an explicit free joint with which the body is connected to the world, see :meth:`ModelBuilder.add_joint_free`.
+
+    Semi-implicit time integration is a variational integrator that
+    preserves energy, however it not unconditionally stable, and requires a time-step
+    small enough to support the required stiffness and damping forces.
+
+    See: https://en.wikipedia.org/wiki/Semi-implicit_Euler_method
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        integrator = wp.FeatherstoneIntegrator(model)
+
+        # simulation loop
+        for i in range(100):
+            state = integrator.simulate(model, state_in, state_out, dt)
+
+    Note:
+        The :class:`FeatherstoneIntegrator` requires the :class:`Model` to be passed in as a constructor argument.
```

Comment:
We can remove this note since it's the default in Newton now.
Also the example code block and SPDX need to be updated.